packages/jsii-rosetta/lib/snippet-dependencies.ts

```diff
@@ -0,0 +1,245 @@
+import * as cp from 'child_process';
+import * as fastGlob from 'fast-glob';
+import * as fs from 'fs-extra';
+import * as os from 'os';
+import * as path from 'path';
+import * as semver from 'semver';
+import { intersect } from 'semver-intersect';
+
+import { findDependencyDirectory, findUp } from './find-utils';
+import * as logging from './logging';
+import { TypeScriptSnippet, CompilationDependency } from './snippet';
+import { mkDict, formatList } from './util';
+
+/**
+ * Collect the dependencies of a bunch of snippets together in one declaration
+ *
+ * We assume here the dependencies will not conflict.
+ */
+export function collectDependencies(snippets: TypeScriptSnippet[]) {
+  const ret: Record<string, CompilationDependency> = {};
+  for (const snippet of snippets) {
+    for (const [name, source] of Object.entries(snippet.compilationDependencies ?? {})) {
+      ret[name] = resolveConflict(name, source, ret[name]);
+    }
+  }
+  return ret;
+}
+
+function resolveConflict(
+  name: string,
+  a: CompilationDependency,
+  b: CompilationDependency | undefined,
+): CompilationDependency {
+  if (!b) {
+    return a;
+  }
+
+  if (a.type === 'concrete' && b.type === 'concrete') {
+    if (b.resolvedDirectory !== a.resolvedDirectory) {
+      throw new Error(`Dependency conflict: ${name} can be either ${a.resolvedDirectory} or ${b.resolvedDirectory}`);
+    }
+    return a;
+  }
+
+  if (a.type === 'symbolic' && b.type === 'symbolic') {
+    // Intersect the ranges
+    return {
+      type: 'symbolic',
+      versionRange: intersect(a.versionRange, b.versionRange),
+    };
+  }
+
+  if (a.type === 'concrete' && b.type === 'symbolic') {
+    const concreteVersion: string = fs.readJsonSync(path.join(a.resolvedDirectory, 'package.json')).version;
+
+    if (!semver.satisfies(concreteVersion, b.versionRange)) {
+      throw new Error(
+        `Dependency conflict: ${name} expected to match ${b.versionRange} but found ${concreteVersion} at ${a.resolvedDirectory}`,
+      );
+    }
+
+    return a;
+  }
+
+  if (a.type === 'symbolic' && b.type === 'concrete') {
+    // Reverse roles so we fall into the previous case
+    return resolveConflict(name, b, a);
+  }
+
+  throw new Error('Cases should have been exhaustive');
```

Comment:
I kind of like using assertion errors for "this should have been impossible".

```suggestion
  throw new assert.AssertionError('Cases should have been exhaustive');
```

Reply:
The arguments to this thing are making me uncomfortable though. It looks pretty narrowly targeted towards unit test assertions:
![image](https://user-images.githubusercontent.com/524162/149336907-431fb43c-8c8a-4e9a-8e0a-a34e6b13d8ab.png)

https://nodejs.org/api/assert.html#new-assertassertionerroroptions